### ISSUE TITLE
Merge some/all of the zig improvements by liyu1981

### ISSUE THREAD

**Issue by SimonKagstrom** [SimonKagstrom/kcov#423]
I.e., this fork:

https://github.com/liyu1981/kcov

@liyu1981, some input?

**Comment by liyu1981** [comment]
@SimonKagstrom I am glad that you find this work.

Currently the implementation of my hack is based on string comparison as from the semantics level I do not find good ways of doing that. I am not sure whether this is something `kcov` prefer, but I can prepare it in a pull request for discussion.

**Comment by SimonKagstrom** [comment]
I think the string comparison is fine. I guess probably other tools would do it that way as well.

The only real comment I have (see the PR) is that I think the comparison can be done in reporter.cc instead, and not in each writer, but feel free to try that out!

Thanks!

**Comment by PasVegan** [comment]
any update ? I see that the PR isn't merged.

**Comment by SimonKagstrom** [comment]
I think it's up to @liyu1981 to investigate and maybe fix what's left

**Comment by mnemnion** [comment]
Heavy Zig + kcov user here, dropping by the issue board trying to figure out fugitive thread stuff. 

Just to say, I build a kcov step like this:

```zig
    const run_kcov = b.addSystemCommand(&.{
        "kcov",
        "--clean",
        "--exclude-line=unreachable,expect(false),panic(,no-coverage",
    });
    run_kcov.setCwd(b.path("."));
    run_kcov.addPrefixedDirectoryArg("--include-pattern=", b.path("src"));
    const coverage_output = run_kcov.addOutputDirectoryArg(".");
    run_kcov.addArtifactArg(module_unit_tests);

    run_kcov.enableTestRunnerMode();

    const install_coverage = b.addInstallDirectory(.{
        .source_dir = coverage_output,
        .install_dir = .{ .custom = "coverage" },
        .install_subdir = "",
    });

    const coverage_step = b.step("coverage", "Generate coverage (kcov must be installed)");
    coverage_step.dependOn(&install_coverage.step);
```
While it would be nice, I suppose, to 'lift' the `--exclude-line` into kcov itself, I would not personally notice if it happened.  

This just seemed like a good place to say thanks, Simon, for your excellent work, which has spared me many hassles down the road! That and leave a little Zig kcov recipe where others might find it.

But I do have one possibly-useful observation: Zig will be adding the ability to test panics once panic-recover lands on master, so maybe not a long-term good candidate for an auto-ignore.  Well, two: note that mine is spelled `panic(`, not `@panic`, because the builtin is not the only way to produce one, it's a whole format-able and otherwise customizable thing.

**Comment by SimonKagstrom** [comment]
Good suggestions @mnemnion ! There's another issue, #489, for regexes  to `--exclude-line`. Not sure if that would help Zig, but perhaps. A set of standard patterns to exclude per file type would also be possible.

The only improvement too your build step I can see is to use `--include-path` instead of `--include-pattern`, since you use a path there anyway.


And thanks for the thanks :-). I don't work very much on kcov anymore, and I'm no longer proud of the code base. Unfortunately, work and more fun projects get in the way. I'll try to keep it alive though!
